app/waf/src/controllers/adc.controller.ts

```diff
@@ -588,3 +580,132 @@ type AddonReqValues = {
   userToken: string;
   tenantId: string;
 };
+
+class AdcStateCtrlr {
+  private defaultCheck = (): Promise<boolean> => {
+    return Promise.resolve(true);
+  };
+  private states: AdcStateEntry[] = [
+    {
+      state: AdcState.NONE,
+      check: Promise.resolve(true),
+      next: [AdcState.POWERON],
+    },
+    {
+      state: AdcState.POWERON,
+      check: this.accessible(),
+      next: [AdcState.ONBOARDED, AdcState.RECLAIMED],
+    },
+    {
+      state: AdcState.ONBOARDED,
+      check: this.onboarded(),
+      next: [AdcState.TRUSTED, AdcState.RECLAIMED],
+    },
+    {
+      state: AdcState.TRUSTED,
+      check: this.trusted(),
+      next: [AdcState.RECLAIMED],
+    },
+    {
+      state: AdcState.RECLAIMED,
+      check: Promise.resolve(this.reclaimed()),
+      next: [AdcState.POWERON],
+    },
+  ];
+
+  constructor(private adc: Adc, private addon: AddonReqValues) {}
+
+  async readyTo(state: string): Promise<boolean> {
+    let stateEntry = this.getStateEntry(this.adc.status);
+    return (await stateEntry.check) && stateEntry.next.includes(state);
+  }
+
+  async gotTo(state: string): Promise<boolean> {
+    return this.getStateEntry(state).check;
+  }
+
+  private getStateEntry(name: string): AdcStateEntry {
+    return this.states.find(s => {
+      return s.state === name;
+    })!;
+  }
+
+  private reclaimed(): boolean {
+    if (this.adc.management) return false;
+    if (this.adc.compute.vmId) return false;
+    for (let net of Object.keys(this.adc.networks)) {
+      if (this.adc.networks[net].portId) return false;
+    }
+    return true;
+  }
+
+  private async trusted(): Promise<boolean> {
+    return Promise.resolve(true);
+  }
+
+  private async onboarded(): Promise<boolean> {
+    let bigipMgr = await this.getBigipMgr();
+    return Promise.all([
+      bigipMgr.getHostname(),
+      bigipMgr.getLicense(),
+      //bigipMgr.getVlans(),
+      //bigipMgr.getSelfips(),
+    ]).then(([hostname, license /*vlans, selfs*/]) => {
+      return (
+        hostname.includes(this.adc.id) && license.registrationKey !== 'none' //&&
+        // Object.keys(vlans).length !== 0 &&
+        // Object.keys(selfs).length !== 0
+      );
+    });
+  }
+
+  private async accessible(): Promise<boolean> {
+    let bigipMgr = await this.getBigipMgr();
+
+    return bigipMgr
+      .getSys()
+      .then(() => Promise.resolve(true))
+      .catch(() => Promise.reject(false));
+  }
+
+  private async getBigipMgr(): Promise<BigIpManager> {
+    if (!this.adc.management)
+      throw new Error(
+        `The management session of ADC is empty, cannot initialize bigip manager.`,
+      );
+
+    let mgmt = this.adc.management;
+    return BigIpManager.instanlize({
+      username: mgmt.username,
+      password: mgmt.password,
+      ipAddr: mgmt.ipAddress,
+      port: mgmt.tcpPort,
+    });
+  }
+}
+
+type AdcStateEntry = {
+  state: string;
+  check: Promise<boolean>;
+  next: string[];
+};
+
+enum AdcState {
```

Comment:
Should we add a new state, eg. 'ACTIVE'? After  TRUSTED, we can install AS3, then ADC can move to ACTIVE state, which indicates this device is ready to process AS3 request.

Reply:
I am not sure if 'NEW' is better than 'NONE' to indicate that state. Maybe 'NONE' make someone a little confused...

Reply:
Yes, we should have a final state `ACTIVE` after all the works are done. 
In the short term, we may have extension installation after trust. 
I suggest to sort out the states incrementally. That's to add such works in later PRs.

Reply:
3ks